# Commençons par importer et observer les données

On importe les librairies magiques qui vont bien

In [15]:
import pandas as pd
import numpy as np

In [16]:
dfTrain = pd.read_csv('train.csv', header = 0)

In [17]:
dfTrain.head(100)

PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
12           13         0       3   
13           14         0       3   
14           15         0       3   
15           16         1       2   
16           17         0       3   
17           18         1       2   
18           19         0       3   
19           20         1       3   
20           21         0       2   
21           22         1       2   
22           23         1       3   
23           24         1       1   
24           25         0       3   
25           26         1       3   
26           27         0       3   
27           28         0       1   
28           29         1       3   
29           30         0       3   
..          ...       ...     ...   
70           71         0       2   
71           72         0       3   
72           73         0       2   
73           74         0       3   
74           75         1       3   
75           76         0       3   
76           77         0       3   
77           78         0       3   
78           79         1       2   
79           80         1       3   
80           81         0       3   
81           82         1       3   
82           83         1       3   
83           84         0       1   
84           85         1       2   
85           86         1       3   
86           87         0       3   
87           88         0       3   
88           89         1       1   
89           90         0       3   
90           91         0       3   
91           92         0       3   
92           93         0       1   
93           94         0       3   
94           95         0       3   
95           96         0       3   
96           97         0       1   
97           98         1       1   
98           99         1       2   
99          100         0       2   

                                                 Name     Sex    Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.00      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.00      1   
2                              Heikkinen, Miss. Laina  female  26.00      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.00      1   
4                            Allen, Mr. William Henry    male  35.00      0   
5                                    Moran, Mr. James    male    NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.00      0   
7                      Palsson, Master. Gosta Leonard    male   2.00      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.00      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.00      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.00      1   
11                           Bonnell, Miss. Elizabeth  female  58.00      0   
12                     Saundercock, Mr. William Henry    male  20.00      0   
13                        Andersson, Mr. Anders Johan    male  39.00      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.00      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.00      0   
16                               Rice, Master. Eugene    male   2.00      4   
17                       Williams, Mr. Charles Eugene    male    NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.00      1   
19                            Masselmani, Mrs. Fatima  female    NaN      0   
20                               Fynney, Mr. Josep

In [4]:
dfTrain.all()

PassengerId     True
Survived       False
Pclass          True
Name            True
Sex             True
Age             True
SibSp          False
Parch          False
Ticket          True
Fare           False
Cabin           True
Embarked        True
dtype: bool

# Explorons quelques méthodes de Pandas

In [18]:
dfTrain.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [19]:
dfTrain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


Nous avons donc un tableau de données de 891 entrées, de 12 colonnes toutes détaillées sur leur type et l'espace occupé par le fichier. 
Nous constatons aussi que beaucoup de cellules ne sont pas remplies. Il va donc nous falloire combler ces vides. 

En cherchant à décrire le tableau celui-ci génère une erreur. Observons ces premières données statistiques pour lesquels nous 
n'avons fait encore aucun calcul. 

In [20]:
dfTrain.describe()

E:\Anaconda\lib\site-packages\numpy\lib\function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000         NaN    0.000000   
50%     446.000000    0.000000    3.000000         NaN    0.000000   
75%     668.500000    1.000000    3.000000         NaN    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

Avec cette fonction Panda a :
- conté le nombre de données renseignées
- calculé pour chaque colonne la moyenne (ce qui n'a pas forcément de sens) et les quartiles
- std standart déviation : écart type par rapport à la moyenne. 
Nous constatons qu'il n'a pas pu le faire pour l'age. En effet il manque presque 20% des données. 
Nous allons devoir nettoyer les données

# Nettoyage des données


En observant les données nous constatons que des âges ne sont pas renseignés. Nous allons donc créer une colonne de copie de l'âge puis nous remplacerons les valeurs par la médiane. 


In [22]:
dfTrain.Age.mean()
dfTrain[['Sex', 'Pclass', 'Age']]
dfTrain[dfTrain['Age'] > 60]
dfTrain[dfTrain['Age'] > 60][['Sex', 'Pclass', 'Age', 'Survived']]
dfTrain[dfTrain['Age'].isnull()][['Sex', 'Pclass', 'Age']]

Sex  Pclass  Age
5      male       3  NaN
17     male       2  NaN
19   female       3  NaN
26     male       3  NaN
28   female       3  NaN
29     male       3  NaN
31   female       1  NaN
32   female       3  NaN
36     male       3  NaN
42     male       3  NaN
45     male       3  NaN
46     male       3  NaN
47   female       3  NaN
48     male       3  NaN
55     male       1  NaN
64     male       1  NaN
65     male       3  NaN
76     male       3  NaN
77     male       3  NaN
82   female       3  NaN
87     male       3  NaN
95     male       3  NaN
101    male       3  NaN
107    male       3  NaN
109  female       3  NaN
121    male       3  NaN
126    male       3  NaN
128  female       3  NaN
140  female       3  NaN
154    male       3  NaN
..      ...     ...  ...
718    male       3  NaN
727  female       3  NaN
732    male       2  NaN
738    male       3  NaN
739    male       3  NaN
740    male       1  NaN
760    male       3  NaN
766    male       1  NaN
768    male       3  NaN
773    male       3  NaN
776    male       3  NaN
778    male       3  NaN
783    male       3  NaN
790    male       3  NaN
792  female       3  NaN
793    male       1  NaN
815    male       1  NaN
825    male       3  NaN
826    male       3  NaN
828    male       3  NaN
832    male       3  NaN
837    male       3  NaN
839    male       1  NaN
846    male       3  NaN
849  female       1  NaN
859    male       3  NaN
863  female       3  NaN
868    male       3  NaN
878    male       3  NaN
888  female       3  NaN

[177 rows x 3 columns]

In [23]:
dfTrain.Age.mean()

29.69911764705882

In [24]:
dfTrain.Age.median()

28.0

L'age médian est de 28 ans alors que la moyenne est de 29.69
On définit une colonne genre. 
On calcule la médiane par classe et par genre

In [30]:
median_ages = np.zeros((2,3))
for i in range(0, 2):
    for j in range(0, 3):
        median_ages[i,j] = dfTrain[(dfTrain['Gender'] == i) & \
                              (dfTrain['Pclass'] == j+1)]['Age'].dropna().median()
median_ages

array([[ 35. ,  28. ,  21.5],
       [ 40. ,  30. ,  25. ]])

In [28]:
# On crée une colonne genre
dfTrain['Gender'] = dfTrain['Sex'].map( {'female': 0, 'male': 1} ).astype(int)
# On fait ensuite une copie de la colonne age:
dfTrain['AgeFill'] = dfTrain['Age']

dfTrain[ dfTrain['Age'].isnull() ][['Gender','Pclass','Age','AgeFill']].head(10)

Gender  Pclass  Age  AgeFill
5        1       3  NaN      NaN
17       1       2  NaN      NaN
19       0       3  NaN      NaN
26       1       3  NaN      NaN
28       0       3  NaN      NaN
29       1       3  NaN      NaN
31       0       1  NaN      NaN
32       0       3  NaN      NaN
36       1       3  NaN      NaN
42       1       3  NaN      NaN

On remplit la colonne des âges en la remplaçant par la médiane

In [31]:
# Et puis on le rempli :
for i in range(0, 2):
    for j in range(0, 3):
        dfTrain.loc[ (dfTrain.Age.isnull()) & (dfTrain.Gender == i) & 
                (dfTrain.Pclass == j+1),\
                'AgeFill'] = median_ages[i,j]
dfTrain[ dfTrain['Age'].isnull() ][['Gender','Pclass','Age','AgeFill']].head(10)

Gender  Pclass  Age  AgeFill
5        1       3  NaN     25.0
17       1       2  NaN     30.0
19       0       3  NaN     21.5
26       1       3  NaN     25.0
28       0       3  NaN     21.5
29       1       3  NaN     25.0
31       0       1  NaN     35.0
32       0       3  NaN     21.5
36       1       3  NaN     25.0
42       1       3  NaN     25.0

In [35]:
dfTrain['AgeIsNull'] = pd.isnull(dfTrain.Age).astype(int)
dfTrain.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  AgeFill  Gender  AgeIsNull  
0      0         A/5 21171   7.2500   NaN        S     22.0       1          0  
1      0          PC 17599  71.2833   C85        C     38.0       0          0  
2      0  STON/O2. 3101282   7.9250   NaN        S     26.0       0          0  
3      0            113803  53.1000  C123        S     35.0       0          0  
4      0            373450   8.0500   NaN        S     35.0       1          0  
5      0            330877   8.4583   NaN        Q     25.0       1          1  
6      0             17463  51.8625   E46        S     54.0       1          0  
7      1            349909  21.0750   NaN        S      2.0       1          0  
8      2            347742  11.1333   NaN        S     27.0       0          0  
9      0            237736  30.0708   NaN        C     14.0       0          0

Nous souhaitons maintenant savoir s'il y a une régression linéaire entre l'âge et la survie :
on importe les librairies


In [38]:
import matplotlib.pyplot as plt
import numpy as np
from sklearn import datasets, linear_model

In [ ]:
Cherchons plutôt si le taux de survie varie avec le prix du billet. 

In [46]:
age = dfTrain.AgeFill().astype(int)


TypeError: 'Series' object is not callable